compliance/collection/main.go

```diff
@@ -154,6 +218,34 @@ func initialClientAndConfig(ctx context.Context, cli sensor.ComplianceServiceCli
 	return client, config, nil
 }
 
+// initializeSendingStream returns a client for sending the data to sensor. It ensures that the stream is open and in case not,
+// then the connection is being reopened with exp. back off
+func initializeSendingStream(ctx context.Context, cli sensor.ComplianceServiceClient) (sensor.ComplianceService_CommunicateClient, error) {
```

Comment:
Few stupid questions from me to make your life seem harder:
What's the difference between `initializeStream` and `initializeSendingStream`?
Does the stream returned by `initializeStream` not allow sending?
Can you please eliminate copy-pasted code and customize some generic function for both cases (if existence of two functions is justified)?

Reply:
Let me answer those valid (not stupid) questions.

> What's the difference between `initializeStream` and `initializeSendingStream`?

One is for sending the other one is for receiving. Placing the sending code in the receiving loop is difficult as the receiver routine must block. Sensor has two separate loops for this and we applied it here separately as well.
The receiving side requires some additional steps when compared to sending (e.g. the config object, the hello message from sensor), so I found it cleaner to add sending separately and leave the receiving code untouched.
The naming of `initializeStream` could be improved to `initializeRcvStream` - would that be ok?

> Does the stream returned by `initializeStream` not allow sending?

It does allow sending. The problem is different. If the connection is down (e.g., due to sensor restart), we would need to reconnect before sending (this is what we do in `initializeSendingStream`). If we would decide to reuse the receiving client for sending then in case of the reconnection on the receiving side, the client would be invalidated. The receiving side would recreate the client in the `initializeStream` loop and the sending side would need to be notified about this to stop using the old client and start using a new one. We could share the client as it it done in sensor, but this is not trivial. 

In sensor, we also use two loops, but we share the client (using unsafe ptrs inside of `centralclient.CentralConnectionFactory`) - this is possible to apply here, but that would be a quite a big change and would make this PR much larger. We can still consider this to be applied in a follow-up if we want. 

In this PR we took a simplified approach (when compared against the solution used in sensor) and we provide two clients and two loops that take care about the reconnections.
Note that mentioning two clients here, I mean two child `sensor.ComplianceService_CommunicateClient`s while still using a single parent `sensor.ComplianceServiceClient`. While I am not fully sure what the precise overhead of this combination is, I think that we still use the same stream and the overhead is minimal. 

> Can you please eliminate copy-pasted code and customize some generic function for both cases (if existence of two functions is justified)?

Yes, this is something we can do.

Reply:
I have removed now the `initializeSendingStream` to address the code duplication issue. We will use `initializeStream` for both cases and have a followup ticket to handle it the same way as in sensor (if we identify that there are benefits coming from this approach).

Reply:
I feel like I'm not able to continue and conclude my review in a timely manner due to many other obligations. At the same time I don't want to hold you back. Please proceed without me, but please make sure you understand how things work and you're 100% confident that the original compliance isn't broken.

Reply:
I feel pretty confident, that we didn't break the logic that was there before. The only thing that we made a bit worse is the introduction of the second client that reinitializes its own connection before sending. Obviously, we need that client, and this client needs to make sure that the connection is up before sending, so let's see how large is the overhead from it.

Maybe we should spin a long-running cluster with our FF enabled to monitor the system with our changes while we work on that epic @Maddosaurus ?

Reply:
@vikin91 A long running cluster would be a great idea. How about I spin one up that runs over the weekend and we merge on Monday morning, given that the cluster and the tests look good?

Reply:
I would say, we first merge ant then do the cluster :) Our changes are behind a FF, so there should be no impact to the rest of the product.